{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "46fca7ed-569e-4ac2-a749-4983452bdf69",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "from io import StringIO\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3de0f786-7e93-401b-a646-22515572df32",
   "metadata": {},
   "source": [
    "Initially the reference gene names were obtained by extracting IDs from the assembly_2 FASTA file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71e75e25-aa29-49d9-82ca-287265c89099",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_gene_names(fasta_file):\n",
    "    gene_names = []\n",
    "\n",
    "    for sequence in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        \n",
    "        # Assuming gene name is the second word in the description line\n",
    "        gene_name = sequence.description.split()[1]\n",
    "        \n",
    "        # If gene We first extracted all the reference gene names from the assembly_2 fasta file. name exists, add it to the list\n",
    "        if gene_name:\n",
    "            gene_names.append(gene_name)\n",
    "\n",
    "    return gene_names\n",
    "\n",
    "fasta_file_path = 'data/assembly_2.prot.fa'\n",
    "\n",
    "# Extract gene names from the FASTA file\n",
    "ref_gene_names = extract_gene_names(fasta_file_path)\n",
    "\n",
    "# Write gene names to a text file\n",
    "with open(output_file_path, 'w') as output_file:\n",
    "    for gene_name in ref_gene_names:\n",
    "        output_file.write(f'{gene_name}\\n')\n",
    "\n",
    "output_file_path = 'ref_gene_names.txt'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "747f1688-55cb-4d55-bbef-bebb5d627e2c",
   "metadata": {},
   "source": [
    "Then a connection was established to the UniProt REST API in order to retrieve protein data for further analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "95f04126-6fb6-4a8a-adfe-813c9ecb842e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "IOPub data rate exceeded.\n",
      "The Jupyter server will temporarily stop sending output\n",
      "to the client in order to avoid crashing it.\n",
      "To change this limit, set the config variable\n",
      "`--ServerApp.iopub_data_rate_limit`.\n",
      "\n",
      "Current values:\n",
      "ServerApp.iopub_data_rate_limit=1000000.0 (bytes/sec)\n",
      "ServerApp.rate_limit_window=3.0 (secs)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# UniProt API URL\n",
    "#url = \"https://rest.uniprot.org/uniprotkb/stream?compressed=true&format=fasta&query=%28brassica+napus%29\"\n",
    "# Send the request to the UniProt API\n",
    "#response = requests.get(url)\n",
    "# Check if the request was successful (status code 200)\n",
    "#if response.status_code == 200:\n",
    "    # Process the API response (e.g., print the content)\n",
    "    #print(response.text)\n",
    "#else:\n",
    "    # Handle errors\n",
    "    #print(f\"Error: {response.status_code}\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f3e70ad-3fc3-4deb-8230-4c06e962e680",
   "metadata": {},
   "source": [
    "The data was imported from the UniProt server with the query as 'organism:Brassica napus'. All the UniProtkb protein entries were written to a TSV file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "11b77628-ba06-47d9-8853-20a2fc6f27d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 / 145683\n",
      "1000 / 145683\n",
      "1500 / 145683\n",
      "2000 / 145683\n",
      "2500 / 145683\n",
      "3000 / 145683\n",
      "3500 / 145683\n",
      "4000 / 145683\n",
      "4500 / 145683\n",
      "5000 / 145683\n",
      "5500 / 145683\n",
      "6000 / 145683\n",
      "6500 / 145683\n",
      "7000 / 145683\n",
      "7500 / 145683\n",
      "8000 / 145683\n",
      "8500 / 145683\n",
      "9000 / 145683\n",
      "9500 / 145683\n",
      "10000 / 145683\n",
      "10500 / 145683\n",
      "11000 / 145683\n",
      "11500 / 145683\n",
      "12000 / 145683\n",
      "12500 / 145683\n",
      "13000 / 145683\n",
      "13500 / 145683\n",
      "14000 / 145683\n",
      "14500 / 145683\n",
      "15000 / 145683\n",
      "15500 / 145683\n",
      "16000 / 145683\n",
      "16500 / 145683\n",
      "17000 / 145683\n",
      "17500 / 145683\n",
      "18000 / 145683\n",
      "18500 / 145683\n",
      "19000 / 145683\n",
      "19500 / 145683\n",
      "20000 / 145683\n",
      "20500 / 145683\n",
      "21000 / 145683\n",
      "21500 / 145683\n",
      "22000 / 145683\n",
      "22500 / 145683\n",
      "23000 / 145683\n",
      "23500 / 145683\n",
      "24000 / 145683\n",
      "24500 / 145683\n",
      "25000 / 145683\n",
      "25500 / 145683\n",
      "26000 / 145683\n",
      "26500 / 145683\n",
      "27000 / 145683\n",
      "27500 / 145683\n",
      "28000 / 145683\n",
      "28500 / 145683\n",
      "29000 / 145683\n",
      "29500 / 145683\n",
      "30000 / 145683\n",
      "30500 / 145683\n",
      "31000 / 145683\n",
      "31500 / 145683\n",
      "32000 / 145683\n",
      "32500 / 145683\n",
      "33000 / 145683\n",
      "33500 / 145683\n",
      "34000 / 145683\n",
      "34500 / 145683\n",
      "35000 / 145683\n",
      "35500 / 145683\n",
      "36000 / 145683\n",
      "36500 / 145683\n",
      "37000 / 145683\n",
      "37500 / 145683\n",
      "38000 / 145683\n",
      "38500 / 145683\n",
      "39000 / 145683\n",
      "39500 / 145683\n",
      "40000 / 145683\n",
      "40500 / 145683\n",
      "41000 / 145683\n",
      "41500 / 145683\n",
      "42000 / 145683\n",
      "42500 / 145683\n",
      "43000 / 145683\n",
      "43500 / 145683\n",
      "44000 / 145683\n",
      "44500 / 145683\n",
      "45000 / 145683\n",
      "45500 / 145683\n",
      "46000 / 145683\n",
      "46500 / 145683\n",
      "47000 / 145683\n",
      "47500 / 145683\n",
      "48000 / 145683\n",
      "48500 / 145683\n",
      "49000 / 145683\n",
      "49500 / 145683\n",
      "50000 / 145683\n",
      "50500 / 145683\n",
      "51000 / 145683\n",
      "51500 / 145683\n",
      "52000 / 145683\n",
      "52500 / 145683\n",
      "53000 / 145683\n",
      "53500 / 145683\n",
      "54000 / 145683\n",
      "54500 / 145683\n",
      "55000 / 145683\n",
      "55500 / 145683\n",
      "56000 / 145683\n",
      "56500 / 145683\n",
      "57000 / 145683\n",
      "57500 / 145683\n",
      "58000 / 145683\n",
      "58500 / 145683\n",
      "59000 / 145683\n",
      "59500 / 145683\n",
      "60000 / 145683\n",
      "60500 / 145683\n",
      "61000 / 145683\n",
      "61500 / 145683\n",
      "62000 / 145683\n",
      "62500 / 145683\n",
      "63000 / 145683\n",
      "63500 / 145683\n",
      "64000 / 145683\n",
      "64500 / 145683\n",
      "65000 / 145683\n",
      "65500 / 145683\n",
      "66000 / 145683\n",
      "66500 / 145683\n",
      "67000 / 145683\n",
      "67500 / 145683\n",
      "68000 / 145683\n",
      "68500 / 145683\n",
      "69000 / 145683\n",
      "69500 / 145683\n",
      "70000 / 145683\n",
      "70500 / 145683\n",
      "71000 / 145683\n",
      "71500 / 145683\n",
      "72000 / 145683\n",
      "72500 / 145683\n",
      "73000 / 145683\n",
      "73500 / 145683\n",
      "74000 / 145683\n",
      "74500 / 145683\n",
      "75000 / 145683\n",
      "75500 / 145683\n",
      "76000 / 145683\n",
      "76500 / 145683\n",
      "77000 / 145683\n",
      "77500 / 145683\n",
      "78000 / 145683\n",
      "78500 / 145683\n",
      "79000 / 145683\n",
      "79500 / 145683\n",
      "80000 / 145683\n",
      "80500 / 145683\n",
      "81000 / 145683\n",
      "81500 / 145683\n",
      "82000 / 145683\n",
      "82500 / 145683\n",
      "83000 / 145683\n",
      "83500 / 145683\n",
      "84000 / 145683\n",
      "84500 / 145683\n",
      "85000 / 145683\n",
      "85500 / 145683\n",
      "86000 / 145683\n",
      "86500 / 145683\n",
      "87000 / 145683\n",
      "87500 / 145683\n",
      "88000 / 145683\n",
      "88500 / 145683\n",
      "89000 / 145683\n",
      "89500 / 145683\n",
      "90000 / 145683\n",
      "90500 / 145683\n",
      "91000 / 145683\n",
      "91500 / 145683\n",
      "92000 / 145683\n",
      "92500 / 145683\n",
      "93000 / 145683\n",
      "93500 / 145683\n",
      "94000 / 145683\n",
      "94500 / 145683\n",
      "95000 / 145683\n",
      "95500 / 145683\n",
      "96000 / 145683\n",
      "96500 / 145683\n",
      "97000 / 145683\n",
      "97500 / 145683\n",
      "98000 / 145683\n",
      "98500 / 145683\n",
      "99000 / 145683\n",
      "99500 / 145683\n",
      "100000 / 145683\n",
      "100500 / 145683\n",
      "101000 / 145683\n",
      "101500 / 145683\n",
      "102000 / 145683\n",
      "102500 / 145683\n",
      "103000 / 145683\n",
      "103500 / 145683\n",
      "104000 / 145683\n",
      "104500 / 145683\n",
      "105000 / 145683\n",
      "105500 / 145683\n",
      "106000 / 145683\n",
      "106500 / 145683\n",
      "107000 / 145683\n",
      "107500 / 145683\n",
      "108000 / 145683\n",
      "108500 / 145683\n",
      "109000 / 145683\n",
      "109500 / 145683\n",
      "110000 / 145683\n",
      "110500 / 145683\n",
      "111000 / 145683\n",
      "111500 / 145683\n",
      "112000 / 145683\n",
      "112500 / 145683\n",
      "113000 / 145683\n",
      "113500 / 145683\n",
      "114000 / 145683\n",
      "114500 / 145683\n",
      "115000 / 145683\n",
      "115500 / 145683\n",
      "116000 / 145683\n",
      "116500 / 145683\n",
      "117000 / 145683\n",
      "117500 / 145683\n",
      "118000 / 145683\n",
      "118500 / 145683\n",
      "119000 / 145683\n",
      "119500 / 145683\n",
      "120000 / 145683\n",
      "120500 / 145683\n",
      "121000 / 145683\n",
      "121500 / 145683\n",
      "122000 / 145683\n",
      "122500 / 145683\n",
      "123000 / 145683\n",
      "123500 / 145683\n",
      "124000 / 145683\n",
      "124500 / 145683\n",
      "125000 / 145683\n",
      "125500 / 145683\n",
      "126000 / 145683\n",
      "126500 / 145683\n",
      "127000 / 145683\n",
      "127500 / 145683\n",
      "128000 / 145683\n",
      "128500 / 145683\n",
      "129000 / 145683\n",
      "129500 / 145683\n",
      "130000 / 145683\n",
      "130500 / 145683\n",
      "131000 / 145683\n",
      "131500 / 145683\n",
      "132000 / 145683\n",
      "132500 / 145683\n",
      "133000 / 145683\n",
      "133500 / 145683\n",
      "134000 / 145683\n",
      "134500 / 145683\n",
      "135000 / 145683\n",
      "135500 / 145683\n",
      "136000 / 145683\n",
      "136500 / 145683\n",
      "137000 / 145683\n",
      "137500 / 145683\n",
      "138000 / 145683\n",
      "138500 / 145683\n",
      "139000 / 145683\n",
      "139500 / 145683\n",
      "140000 / 145683\n",
      "140500 / 145683\n",
      "141000 / 145683\n",
      "141500 / 145683\n",
      "142000 / 145683\n",
      "142500 / 145683\n",
      "143000 / 145683\n",
      "143500 / 145683\n",
      "144000 / 145683\n",
      "144500 / 145683\n",
      "145000 / 145683\n",
      "145500 / 145683\n",
      "145683 / 145683\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import requests\n",
    "from requests.adapters import HTTPAdapter, Retry\n",
    "\n",
    "re_next_link = re.compile(r'<(.+)>; rel=\"next\"')\n",
    "retries = Retry(total=5, backoff_factor=0.25, status_forcelist=[500, 502, 503, 504])\n",
    "session = requests.Session()\n",
    "session.mount(\"https://\", HTTPAdapter(max_retries=retries))\n",
    "\n",
    "def get_next_link(headers):\n",
    "    if \"Link\" in headers:\n",
    "        match = re_next_link.match(headers[\"Link\"])\n",
    "        if match:\n",
    "            return match.group(1)\n",
    "\n",
    "def get_batch(batch_url):\n",
    "    while batch_url:\n",
    "        response = session.get(batch_url)\n",
    "        response.raise_for_status()\n",
    "        total = response.headers[\"x-total-results\"]\n",
    "        yield response, total\n",
    "        batch_url = get_next_link(response.headers)\n",
    "\n",
    "\n",
    "url = 'https://rest.uniprot.org/uniprotkb/search?fields=accession%2Cprotein_name%2Cgene_names&format=tsv&query=%28brassica+napus%29&size=500'\n",
    "progress = 0\n",
    "\n",
    "with open('brassica_napus.tsv', 'w') as f:\n",
    "    for batch, total in get_batch(url):\n",
    "        lines = batch.text.splitlines()\n",
    "        if not progress:\n",
    "            print(lines[0], file=f)\n",
    "        for line in lines[1:]:\n",
    "            print(line, file=f)\n",
    "        progress += len(lines[1:])\n",
    "        print(f'{progress} / {total}')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3658022-6318-4fda-8ad2-c9f2fa99bd3d",
   "metadata": {},
   "source": [
    "All standard gene identifiers were subsequently mapped to their corresponding UniProt accession identifiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "db9399c3-6244-4dce-91f4-0027ed86c69c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_tsv(input_tsv_path, output_tsv_path, keywords):\n",
    "    # Read the TSV file into a dataframe\n",
    "    df = pd.read_csv(input_tsv_path, sep='\\t')\n",
    "\n",
    "    # Filter rows containing any of the keywords (case-insensitive)\n",
    "    filtered_df = df[df['Gene Names'].str.contains('|'.join(keywords), case=False, na=False)]\n",
    "\n",
    "    # Save the filtered dataframe to a new TSV file\n",
    "    filtered_df.to_csv(output_tsv_path, sep='\\t', index=False)\n",
    "\n",
    "input_file_path = 'brassica_napus.tsv'\n",
    "output_file_path = 'matched_output.tsv'\n",
    "keywords_to_search = ref_gene_names\n",
    "\n",
    "filter_tsv(input_file_path, output_file_path, keywords_to_search)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab48f2f3-a5e2-4d2c-b54d-c18e9d3ff4e5",
   "metadata": {},
   "source": [
    "For the second half of task 2, our aim was to study enyzmatic activity of Hydrolase proteins on Ester bonds, so the query made to the API used the parameters: 'organism: Brassica napus' and Enzyme Activity(ec:3.1.-.-). The data was similarly written to a TSV file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "855d15c4-3f2b-4bc5-bda9-6c0e6d70bcc9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "500 / 1462\n",
      "1000 / 1462\n",
      "1462 / 1462\n"
     ]
    }
   ],
   "source": [
    "url2 = \"https://rest.uniprot.org/uniprotkb/search?fields=accession%2Cprotein_name%2Cgene_names&format=tsv&query=%28%28brassica+napus%29+AND+%28ec%3A3.1.-.-%29%29&size=500\"\n",
    "progress = 0\n",
    "\n",
    "with open('enzyme_activity.tsv', 'w') as f:\n",
    "    for batch, total in get_batch(url2):\n",
    "        lines = batch.text.splitlines()\n",
    "        if not progress:\n",
    "            print(lines[0], file=f)\n",
    "        for line in lines[1:]:\n",
    "            print(line, file=f)\n",
    "        progress += len(lines[1:])\n",
    "        print(f'{progress} / {total}')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c276d77-5695-4299-a405-723a060bb7b3",
   "metadata": {},
   "source": [
    "The genes in specific enzyme activity data were then matched with the genes in aligned sequences from assembly_1 and assembly_2. These proteins will give us an idea of which proteins in assembly_1 are involved in Hydrolase activity on Ester bonds. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "d80308ed-d293-48cf-9e31-ecac0629c39c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "enzyme_file_path = 'enzyme_activity.tsv'\n",
    "fasta_file_path = 'data/output.fasta'\n",
    "output_tsv_path = 'required_proteins.tsv'\n",
    "\n",
    "# Read gene names from enzyme_filtered_file, keeping only genes starting with 'Bna'\n",
    "enzyme_gene_names = [gene_name for gene_name in pd.read_csv('enzyme_activity.tsv', sep='\\t')['Gene Names'].dropna().str.split().explode().tolist() if gene_name.startswith('Bna')]\n",
    "enzyme_gene_names = set(enzyme_gene_names)\n",
    "\n",
    "# Read gene names from output.fasta\n",
    "fasta_gene_names = set(record.description.split()[1] for record in SeqIO.parse(fasta_file_path, \"fasta\"))\n",
    "\n",
    "# Find common gene names\n",
    "common_gene_names = enzyme_gene_names.intersection(fasta_gene_names)\n",
    "\n",
    "# Read enzyme_file and filter entries based on common gene names\n",
    "enzyme_df = pd.read_csv(enzyme_file_path, sep='\\t')\n",
    "filtered_entries = enzyme_df[enzyme_df['Gene Names'].str.contains('|'.join(common_gene_names), case=False, na=False)]\n",
    "\n",
    "# Save the filtered values to a new TSV file containing the required proteins\n",
    "filtered_df.to_csv(output_tsv_path, sep='\\t', index=False)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
